### ISSUE TITLE
Train model on synthetic data, agnostic to contrast

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/axondeepseg#591]
The idea would be to generate synthetic data from real data + ground truth, and use approaches such as GANs. The goal would be to generate images with infinite amount of contrasts in order to have models become agnostic to image contrast. 

The idea is similar to what has been proposed in MRI: https://github.com/BBillot/SynthSeg

**Comment by hermancollin** [comment]
I've been looking into the SynthSeg project and I think it could work for ADS. The way they generate synthetic MRI training data is by sampling a Gaussian Mixture Model conditioned on a label map for each label. From what I understand, this is how they obtain different contrasts and resolutions within a single training image, which in turn makes the model agnostic to contrast. The domain randomization strategy also enables the model to learn from images that are always different even if they come from the same label maps.

The GAN approach could also work but it would be a different process. For example, a conditional GAN could be trained for image-to-image translation, as described in [this blog post](https://theailearner.com/2019/10/29/image-to-image-translation-using-conditional-gan/), to generate images from existing masks. We could then generate realistic microscopy data from masks and potentially train multiple networks to have multiple contrasts, but this feels more like a data augmentation technique because the training examples would still look somewhat realistic. However, it could also be an advantage because we could multiply the amount of data we have from a given dataset, whereas the SynthSeg method seems to have less impact on the quantity of data. Instead, it creates unrealistic and challenging images that are potentially more efficient for training.

In short, both strategies could yield similar results but I feel like the GAN approach is more "brute-force" and the GMM method more direct.

**Comment by hermancollin** [comment]
This issue hasn't been updated for a while so here's whats been done. Also looping in @Nish228.
- originally, SynthSeg was mentioned as a starting point. Looking deeper into their codebase, I implemented an equivalent version of what they do to generate images on this repo: [gmm-sampled-histology](https://github.com/hermancollin/gmm-sampled-histology). Long story short: this is useless for us. This method produced images too different from real microscopy ones, so once we saw some preliminary results, I stopped looking into it. 
- the GAN approach mentioned in my first comment is much more usable. I tried a few things and got pretty convincing results generating TEM images. For example, this image is fake:
![epoch215_fake_B](https://user-images.githubusercontent.com/83031821/170776359-9d6a4277-05a2-48b7-bd06-ea311fcb4d7c.png)
I also plan on pushing this further but got caught up with other issues.
- in addition to generating microscopy images, I checked for mask generation, i.e. segmentation labels generated by a GAN. I made a rough DC-GAN implementation here for this: [axonmyelin-mask-dcgan](https://github.com/hermancollin/axonmyelin-mask-dcgan). It doesn't work very well yet but this was also something I wished to explore further. The goal would be to generate segmentation masks from "nothing", then translate these masks to microscopy images in different contrasts. Then train a model on this synthetic data, and see what happens.

Another idea that wasn't explicitly discussed here would be to train a model on all the data we have across all modalities. This would be a huge dataset... maybe it could yield a decent contrast-agnostic model. Also, if we explore the synthetic data path further, this would be an excellent baseline to see if the synthetic nature of the data provides any benefit. I'm not sure how the resampling would work... also the model would probably have a slightly different architecture than what we normally use for "single contrast" models.

**Comment by mathieuboudreau** [comment]
This is what @Nish228 has been working on, I'll assign it to her and maybe she can provide a brief update with some links to the repository
